crates/durability/src/lib.rs

```diff
@@ -0,0 +1,101 @@
+use std::sync::Arc;
+
+pub use spacetimedb_commitlog::{error, payload::Txdata, Decoder, Transaction};
+
+mod imp;
+pub use imp::{local, Local};
+
+/// Transaction offset.
+///
+/// The transaction offset is essentially a monotonic counter of all
+/// transactions submitted to the durability layer, starting from zero.
+///
+/// While the implementation may not guarantee that the sequence contains no
+/// gaps, it must guarantee that a higher transaction offset implies durability
+/// of all offsets smaller than it.
+pub type TxOffset = u64;
+
+/// The durability API.
+///
+/// NOTE: This is a preliminary definition, still under consideration.
+///
+/// A durability implementation accepts a payload representing a single database
+/// transaction via [`Durability::append_tx`] in a non-blocking fashion. The
+/// payload _should_ become durable eventually. [`TxOffset`]s reported by
+/// [`Durability::durable_tx_offset`] shall be considered durable to the
+/// extent the implementation can guarantee.
+pub trait Durability: Send + Sync {
+    /// The payload representing a single transaction.
+    type TxData;
+
+    /// Submit the transaction payload to be made durable.
+    ///
+    /// This method must never block, and accept new transactions even if they
+    /// cannot be made durable immediately.
+    ///
+    /// A permanent failure of the durable storage may be signalled by panicking.
+    fn append_tx(&self, tx: Self::TxData);
+
+    /// The [`TxOffset`] considered durable.
+    ///
+    /// A `None` return value indicates that the durable offset is not known,
```

Comment:
Do we want to return `Result<Option<TxOffset>, Error>` to distinguish between these two cases?

Reply:
The thing is that there is no recoverable error -- at all, actually. Either things converge eventually, or the commitlog becomes unwriteable, in which case we panic on the next write. We could try to crash the datastore earlier if that's desired.